tests/select_constraints_tests.cpp

```diff
@@ -0,0 +1,219 @@
+            std::vector<Employee> expected;
+            expected.push_back(paul);
+            expected.push_back(allen);
+            expected.push_back(teddy);
+            expected.push_back(mark);
+            expected.push_back(david);
+            expected.push_back(kim);
+            expected.push_back(james);
+            REQUIRE_THAT(allEmployees, UnorderedEquals(expected));
+        }
+    }
+    SECTION("distinct") {
+        storage.insert(Employee{-1, "Paul", 24, "Houston", 20000.0});
+        storage.insert(Employee{-1, "James", 44, "Norway", 5000.0});
+        storage.insert(Employee{-1, "James", 45, "Texas", 5000.0});
+
+        std::vector<std::string> names;
+        decltype(names) expected;
+        SECTION("without distinct") {
+            SECTION("without prepared statement") {
+                names = storage.select(&Employee::name);
+            }
+            SECTION("with prepared statement") {
+                auto statement = storage.prepare(select(&Employee::name));
+                names = storage.execute(statement);
+            }
+            expected.push_back("Paul");
+            expected.push_back("Allen");
+            expected.push_back("Teddy");
+            expected.push_back("Mark");
+            expected.push_back("David");
+            expected.push_back("Kim");
+            expected.push_back("James");
+            expected.push_back("Paul");
+            expected.push_back("James");
+            expected.push_back("James");
+        }
+        SECTION("with distinct") {
+            SECTION("without prepared statement") {
+                names = storage.select(distinct(&Employee::name));
+            }
+            SECTION("with prepared statement") {
+                auto statement = storage.prepare(select(distinct(&Employee::name)));
+                names = storage.execute(statement);
+            }
+            expected.push_back("Paul");
+            expected.push_back("Allen");
+            expected.push_back("Teddy");
+            expected.push_back("Mark");
+            expected.push_back("David");
+            expected.push_back("Kim");
+            expected.push_back("James");
+        }
+        REQUIRE_THAT(names, UnorderedEquals(expected));
+    }
+#ifdef SQLITE_ORM_OPTIONAL_SUPPORTED
+    SECTION("as_optional") {
+        SECTION("prepared statements bindings") {
+            auto statement = storage.prepare(select(as_optional(5)));
+            auto rows = storage.execute(statement);
+            decltype(rows) expected;
+            expected.push_back(5);
+            REQUIRE(rows == expected);
+
+            get<0>(statement) = 10;
+            expected.clear();
+            expected.push_back(10);
+            rows = storage.execute(statement);
+            REQUIRE(rows == expected);
+        }
+        SECTION("just names") {
+            auto rows = storage.select(as_optional(&Employee::name));
+            decltype(rows) expected;
```

Comment:
I think we should check that `decltype(rows)::value_type` type is `optional` (`static_assert` or something like that) because now test will be passed if `as_optional` doesn't change result type.

Reply:
I thought about it cause you have already wrote it in your example code. But it turned out that it is over-engineering cause we already comparing vectors inside `REQUIRE` macro. It types are different the code will not compile.